libs/prelude/Prelude/Types.idr

```diff
@@ -455,10 +455,17 @@ Foldable List where
 
   foldMap f = foldl (\acc, elem => acc <+> f elem) neutral
 
+-- tail recursive O(n) implementation of `(>>=)` for `List`
+listBind : List a -> (a -> List b) -> List b
+listBind as f = go Nil as
+  where go : List b -> List a -> List b
+        go xs []        = reverse xs
+        go xs (y :: ys) = go (reverseOnto xs (f y)) ys
+
```

Comment:
Should we have  a `%transform` rule turning `concatMap` into `listBind`?

Reply:
I like the idea, but I'm not very fluent with `%transform` rules. Can we specialize such a rule for a constrained function (in case of `concatMap` this would require specializing both `m` and `t` to `List`, yes?). Also, if we used such a transform rule for `foldMap` instead, would this then automatically also apply to `concatMap` and `concat`? Or would all three require their own transform rule?

Reply:
Yes we can cf.

> libs/prelude/Prelude/Types.idr:%transform "fastConcat" concat {t = List} {a = String} = fastConcat

`concatMap` does seem to be defined in terms of `foldMap` so adding that more fundamental `%transform` rule should indeed be enough.

Reply:
OK. Thanks for the clarifications. I'm going to give this a try and add some additional test cases to make sure the transform rules apply for all three functions. But this will have to wait till tonight.

Reply:
So, I need some help with this. I added the transform rules to `Prelude.Types` (a single rule for `foldMap` only seems to be not enough). But unless I also add them to the source file in the `prelude/bind001` test (which I did in the latest commit), they don't seem to apply.

Reply:
I wouldn't expect the `concatMap`  and `concat` ones to be needed as their own bodies are defined in terms of
`foldMap`  and so would be optimised just fine.

Is the fact the rule doesn't fire for `foldMap` due to the fact that the `Foldable` is public export and things get inlined?
Is it possible to attach a `%noinline` to that definition of `foldMap`?

Reply:
I can't annotate the implementation of `foldMap` of `List` with `%noinline`: The parser won't accept it. Also, removing any of the transform rules from the test source file (test/prelude/bind001/bind.idr) will result in the corresponding function not being transformed, so at least there, all of them are needed. The rule for `foldMap` alone does not seem to affect `concatMap` or `concat`.

I also tried annotating the implementation of `Foldable List` with `%noinline` but to no avail.

Reply:
Ok, so my suggestion is to
1. abandon the `%transform` proposal & merge this PR
2. open a bug report about the strange behaviour we are seeing

No need to tie this particular contribution to the bug (?) being resolved.

Reply:
I ditched the transform rules and will open an issue later on.